### DISCUSSION TITLE
Unable to get file operation notifications (willRename, etc)

### DISCUSSION THREAD

**Discussion by GeForceLegend** [ebkalderon/tower-lsp#387] [Q&A] [answered]
I'm tring to implement a language server for Minecraft:Java Edition shaderpacks that require Optifine, and I found that my language server cannot recieve willRename notifications. Source code here: [GeForceLegend/vscode-mcshader/rename_feature](https://github.com/GeForceLegend/vscode-mcshader/tree/rename_feature).

I searched in issues of several vscode related repo and google, and found nothing useful. There are some guys on stackOverflow have the same problem with me, but the answer doesn't helps.
I'v set file operation capabilities and even tried to register it dynamically, but they all get failed.
I even tried add did_create capability with `**/*` pattern, but still cannot see expected error message in log output.
```rust
// will_rename part in server capabilities
will_rename: Some(FileOperationRegistrationOptions {
    filters: vec![FileOperationFilter{
        scheme: Some(String::from("file")),
        pattern: FileOperationPattern {
            glob: String::from("**/*.{vsh,gsh,fsh,csh,glsl}"),
            matches: Some(FileOperationPatternKind::File),
            options: None
        }
    }]
})
// dynamic registration for will rename. it's now commented in source code.
Registration {
    id: String::from("workspace/willRenameFiles"),
    method: String::from("workspace/willRenameFiles"),
    // glsl_file_operation is a FileOperationRegistrationOptions with same filter in above capabilities code
    register_options: Some(serde_json::to_value(&glsl_file_operation).unwrap()),
},
```

**Comment by ebkalderon** [comment, marked as answer]
Hi there, @GeForceLegend! I think your server capabilities code looks fine, on a cursory glance (though I believe `did_rename` needs to be activated also by being set to `Some(_)`, if you'd like to receive that type of notification statically, but that's a minor thing). 

Have you verified that the client extension has `willRename` enabled in its `TextDocumentSyncClientCapabilities` as well? You can find out by inspecting the `params` passed into the `LanguageServer::initialize()` method. I'm currently away from my computer, so I can't confirm right now, unfortunately.

Here's a page you might find helpful regarding dynamic capability registration, if you haven't looked at it already: [link](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#client_registerCapability)

**Reply by GeForceLegend** [reply]
I tried code below and the log says the client supports will rename feature.
```rust
match params.capabilities.workspace.unwrap().file_operations.unwrap().will_rename {
    Some(true) => info!("Client supports will rename feature"),
    _ => info!("Client does not support will rename feature"),
}
```
I'v done same thing as willRename to didRename, but as I said, my server cannot receive any file operation notification, even a didCreate with `**/*` does not call related function. I'm not sure if this is client or server related. After all, thanks for your help! At least I know the client should support it now :D

**Reply by ebkalderon** [reply]
That's good to know! Have you managed to inspect the incoming and outgoing JSON-RPC calls to verify whether a `workspace/willRenameFiles` request or `workspace/didRenameFiles` notification came in from the client, but `tower-lsp` simply didn't execute the handler?

**Reply by GeForceLegend** [reply]
Not tested yet, since I just start learning rust not long ago, I dont know how to it currently, but I'll try it later. Is there any sample or guide for it so I can get on it faster?

**Reply by GeForceLegend** [reply]
I finally find a way: below client code forces send a `workspace/didRenameFiles` notification when renaming a file, and get an expected `WARN Got a workspace/didRenameFiles notification, but it is not implemented` in log output. Now the problem is why the client does not send this notification when renaming files.
```typescript
// A is a class implements RenameFilesParams
// B is a class implements FileRename
this.extension.context.subscriptions.push(...[vscode.workspace.onDidRenameFiles((fileWillRenameEvent) => {
  let a = new A
  log.info(a.files)
  log.info('Send rename notification, content:')
  fileWillRenameEvent.files.forEach((file) => {
    let b = new B(file.oldUri.toString(), file.newUri.toString())
    log.info(b)
    a.files.push(b)
  })
  this.extension.lspClient.sendNotification(lsp.DidRenameFilesNotification.type, a)
  log.info('Send rename notification, content:' + a)
})])
```

**Reply by ebkalderon** [reply]
> Not tested yet, since I just start learning rust not long ago, I dont know how to it currently, but I'll try it later. Is there any sample or guide for it so I can get on it faster?

Unfortunately, there are no samples on hand for this particular part of the LSP spec (`willRename`/`didRename`), but the end behavior of `tower-lsp` should be basically identical to the numerous JS examples available online (if it isn't, that is certainly a bug; tickets and bugfixes are always welcome). It is ultimately the responsibility of the language client implementation to determine when to send these requests/notifications and to correctly interpret the `*` glob patterns received from the language server.

Since it appears the client is indicating support for file operation notifications and `tower-lsp` appears to be routing these notifications correctly whenever they are received, I suspect this is likely a client-side issue. :thinking:

**Reply by GeForceLegend** [reply]
This *should* be automaticly sent by vscode's language client api. Maybe I'll rewrite the client later, great thanks for your help!

**Reply by ebkalderon** [reply]
No problem. Good luck with the rest of your project!
